herbarium-sheet-plant-organ-detection/main.py

```diff
@@ -22,74 +26,89 @@ def start_kafka(predictor: DefaultPredictor) -> None:
+    producer = KafkaProducer(
+        bootstrap_servers=[os.environ.get('KAFKA_PRODUCER_HOST')],
+        value_serializer=lambda m: json.dumps(m).encode('utf-8'))
     for msg in consumer:
         logging.info(msg.value)
         json_value = msg.value
-        image_uri = json_value['data']['ac:accessURI']
-        additional_info_annotations = run_object_detection(image_uri, predictor)
+        image_uri = json_value['object']['digitalEntity']['ac:accessURI']
+        additional_info_annotations, width, height = run_object_detection(image_uri, predictor)
         logging.info('Publishing the result: %s', json_value)
-        annotation = map_to_annotation(json_value, additional_info_annotations)
-        send_updated_opends(annotation, producer)
+        event = map_to_annotation(json_value, additional_info_annotations, width, height)
+        event["jobId"] = json_value['jobId']
+        send_updated_opends(event, producer)
 
 
-def map_to_annotation(json_value, additional_info_annotations) -> dict:
-    values = []
+def map_to_annotation(json_value, additional_info_annotations, width, height) -> dict:
+    annotations = []
     for value in additional_info_annotations:
-        values.append(
-            {
-                'class': value['class'],
-                'score': value['score'],
-                'selector': {
-                    'type': 'FragmentSelector',
-                    'conformsTo': 'http://www.w3.org/TR/media-frags/',
-                    'value': "xywh=" + str(value['boundingBox'][0]) + ',' + str(value['boundingBox'][1]) + ',' + str(
-                        value['boundingBox'][2]) + ',' + str(value['boundingBox'][3])
+        body_values = [{
+            'class': value['class'],
+            'score': value['score'],
+        }]
+        selector = {
+            'oa:selector': {
+                'ods:type': 'FragmentSelector',
+                'dcterms:conformsTo': 'https://www.w3.org/TR/media-frags/',
+                'ac:hasRoi': {
+                    "ac:xFrac": value['boundingBox'][0] / width,
+                    "ac:yFrac": value['boundingBox'][1] / height,
+                    "ac:widthFrac": (value['boundingBox'][2] - value['boundingBox'][0])/width,
+                    "ac:heightFrac": (value['boundingBox'][3] - value['boundingBox'][1]) / height
                 }
             }
-        )
-    annotation = {
-        'type': 'Annotation',
-        'motivation': 'https://hdl.handle.net/pid-machine-enrichment',
-        'creator': 'https://hdl.handle.net/enrichment-service-pid',
-        'created': timestamp_now(),
-        'target': {
-            'id': json_value['id'],
-            'type': 'https://hdl.handle.net/21...'
-        },
-        'body': {
-            'source': json_value['data']['ac:accessURI'],
-            'purpose': 'describing',
-            'values': values
         }
+        annotation = {
+            'rdf:type': 'Annotation',
+            'oa:motivation': 'ods:adding',
```

Comment:
I am missing creator and created

Reply:
Those are handled in the backend/processing service

Reply:
Pretty sure that only generator is covered by the processing service. The creator/created timestamp are responsibility of the MAS. When the annotation hits the processing service it doesn't know from which MAS it comes so it doesn't know who created the annotation. Therefore it needs to be in the MAS. I just made a PR so the MAS name and MAS pid are in env variables of the MAS and can be used (see gbif mas PR)